multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -25,53 +31,130 @@ struct {
 } dma_common[2];
 
 
-enum { dma1 = 0, dma2 };
-
-
 static const int dma2pctl[] = { pctl_dma1, pctl_dma2 };
 
 
-static const struct dma_map {
+struct libdma_per {
 	char dma;
 	char channel[2];
 	char reqmap;
-} spiChanMap[3] = {
+};
+
+
+/* clang-format off */
+enum { dma1 = 0, dma2 };
+
+
+enum { isr = 0, ifcr, cselr = 42 };
+
+
+enum { ccr = 0, cndtr, cpar, cmar };
+/* clang-format on */
+
+
+static const struct libdma_per libdma_persSpi[] = {
 	{ dma1, { 1, 2 }, 0x1 },
 	{ dma1, { 3, 4 }, 0x1 },
 	{ dma2, { 0, 1 }, 0x3 },
 };
 
 
+static const struct libdma_per libdma_persUart[] = {
+	{ dma1, { 4, 3 }, 0x2 },  // or { dma2, { 6, 5 }, 0x2 }
+	{ dma1, { 5, 6 }, 0x2 },
+	{ dma1, { 2, 1 }, 0x2 },
+	{ dma2, { 4, 2 }, 0x2 },
+	{ dma2, { 1, 0 }, 0x2 },
+};
+
+
 static const struct {
 	unsigned int base;
 	int irq_base;
 } dmainfo[2] = { { 0x40020000, 11 }, { 0x40020400, 56 } };
 
 
-enum { isr = 0, ifcr, cselr = 42 };
+static struct {
+	enum {
+		dma_transferNull = 0,
+		dma_transferInf,
+		dma_transferOnce,
+	} type;
+	union {
+		struct {
+			void (*fn)(void *arg, int type);
+			void *arg;
+		} inf;
+		struct {
+			volatile int *done_flag;
+			volatile unsigned int *channel_base;
+		} once;
+	};
+} dma_transfers[2][8];
+
+
+static void libdma_prepareTransfer(volatile unsigned int *channel_base, void *maddr, size_t len, int flags)
+{
+	*(channel_base + cmar) = (unsigned int)maddr;
+	*(channel_base + cndtr) = len;
+	dataBarier();
+	*(channel_base + ccr) |= flags | 0x1;
+	dataBarier();
+}
 
 
-enum { ccr = 0, cndtr, cpar, cmar };
+static void libdma_unprepareTransfer(volatile unsigned int *channel_base)
+{
+	dataBarier();
+	/* Disable interrupts, disable channel */
+	*(channel_base + ccr) &= ~(DMA_TCIE_FLAG | DMA_HTIE_FLAG | 0x1);
+	dataBarier();
+}
 
 
-static int irqHandler(unsigned int n, void *arg)
+static int libdma_irqHandler(unsigned int n, void *arg)
 {
-	int channel = n - dma_common[(int)arg].irq_base - 16;
-	volatile unsigned int *base = dma_common[(int)arg].base;
+	int dma = (int)arg;
+	int channel = n - dma_common[dma].irq_base - 16;
```

Comment:
Oops, it looks like a bug. Interrupt numbers for DMA2 channels are noncontiguous (see `irqnum` in `libdma_configureChannel()`), so the formula here is invalid. It will bite us when alternative UART DMA mapping will be used, so please either delegate this to me or fix it in this PR.

(Note: Maybe get rid of `irq_base` in `dma_common` and `dmainfo` and use a pure function for channel calculation?)

Reply:
good catch thanks